doc/developer/design/20250415_lift_max_result_size_limitation.md

```diff
@@ -0,0 +1,113 @@
+# Lift Max Result Size Limitation
+
+## Context
+
+Currently, results of SELECT queries are sent back from the cluster to
+`environmentd` via the compute protocol and these results are fully
+materialized (stored in memory) in `environmentd` before sending them out to
+the client.
+
+This has several implications:
+
+- Sending large results "clogs up" the cluster/controller communication
+- The amount of memory we want to give to `environmentd` limits the size of
+  results we can return
+
+In practice the above make it so we currently limit the size of results using a
+`max_result_size` parameter, and some larger customers are chafing against
+that.
+
+The overall goal of this design is to lift this limitation, but there is
+nuance, especially when it comes to different query modifiers. For example,
+queries with a ORDER BY and/or LIMIT clause require results from the cluster to
+be combined somewhere, to impose the required order. Currently, this is
+happening in `environmentd`, but could happen on the cluster side. We think it
+is easier to lift result-size limitations for queries that don't require
+post-processing. We should be able to just "stream them through", without
+materializing in `environmentd`.
```

Comment:
FWIW, I think it wouldn't be super hard to support these cases using the `Consolidator` machinery we have now. Happy to chat about this if you're interested!

Reply:
Nice! I was thinking this might be possible, glad to know you think the same!